src/Url.php

```diff
@@ -1066,4 +1066,36 @@ public static function peelUrl($haystack, $prelude_string, $wrapper_start, $wrap
     }
     return FALSE;
   }
+
+  /**
+   * Checks if given URL matches a list of candidates for a default document.
+   *
+   * @param string $url
+   *   The URL to be tested
+   * @param array $candidates
+   *   A list of potential document names that could be indexes.
+   *   Defaults to "default" and "index".
+   *
+   * @return mixed
+   *   string - The base path if a matching document is found.
+   *   bool - FALSE if no matching document is found.
+   */
+  public static function getRedirectIfIndex($url, $candidates = array("default", "index")) {
+    // Assume no match.
+    $root_path = FALSE;
+
+    // Filter through parse_url to separate out querystrings and etc.
+    $path = parse_url($url, PHP_URL_PATH);
+    $filename = strtolower(pathinfo($path, PATHINFO_FILENAME));
+
+    // Test parsed URL
+    if (!empty($filename) && in_array($filename, $candidates)) {
+      $root_path = substr($path, strlen($path) - strlen($candidate));
+      // Re-append the query string and fragment to this path.
+      $root_path .= parse_url($url, PHP_URL_QUERY);
+      $root_path .= parse_url($url, PHP_URL_FRAGMENT);
```

Comment:
`blah1/blah2/index.html#awesome`
would end up looking like 
`blah1/blah2awesome`

For the query and the fragment you really do want to reassemble them like
````
$new_uri = array(
  'path' => $root_path,
'query' => parse_url($url, PHP_URL_QUERY),
'fragment' => parse_url($url, PHP_URL_FRAGMENT),
);

$root_path = self::reassembleURL($new_uri, FALSE)

````